### PR TITLE
Retry durable-nonce transactions

### LINKED ISSUE #10825: SendTransactionService doesn't retry Nonce transactions
#### Problem
When the SendTransactionService processes a Nonce transaction with a nonce that is not in the blockhash cache, it attempts to send the transaction once, then evaluates the last_valid_slot as 0 and drops the transaction from the transactions map (does not retry). Not great ux.

#### Proposed Solution
Add special handling for Nonce transactions in the SendTransactionService such that Nonce transactions are retried until the nonce in the associate nonce account is advanced.

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
When the SendTransactionService processes a Nonce transaction with a nonce that is not in the blockhash cache, it attempts to send the transaction once, then evaluates the last_valid_slot as 0 and drops the transaction from the transactions map (does not retry).

#### Summary of Changes
Retry sending durable-nonce transactions until the nonce in the associated nonce account is advanced or the transaction lands.

Fixes #10825

### CODE REVIEW

File: core/src/send_transaction_service.rs [hunk 1]
```diff
@@ -208,49 +217,58 @@ impl SendTransactionService {
...
                 inc_new_counter_info!("send_transaction_service-rooted", 1);
-                false
+                return false;
+            }
+            if let Some((nonce_pubkey, durable_nonce)) = transaction_info.durable_nonce_info {
+                let nonce_account = working_bank.get_account(&nonce_pubkey).unwrap_or_default();
+                if !nonce_account::verify_nonce_account(&nonce_account, &durable_nonce) {
+                    info!("Dropping expired durable-nonce transaction: {}", signature);
+                    result.expired += 1;
+                    inc_new_counter_info!("send_transaction_service-expired", 1);
```
Comment (CriesofCarrots) [inline core/src/send_transaction_service.rs:227]: Do we care about tracking this case with a separate metric?

Comment (t-nelson) [reply core/src/send_transaction_service.rs:229]: Off hand, I can't think of anything actionable based off such a metric.  Total nonced TXs in the queue might be interesting though.

File: core/src/send_transaction_service.rs [hunk 2]
```diff
@@ -208,49 +217,60 @@ impl SendTransactionService {
                 info!("Transaction is rooted: {}", signature);
                 result.rooted += 1;
                 inc_new_counter_info!("send_transaction_service-rooted", 1);
-                false
+                return false;
+            }
+            if let Some((nonce_pubkey, durable_nonce)) = transaction_info.durable_nonce_info {
+                let nonce_account = working_bank.get_account(&nonce_pubkey).unwrap_or_default();
+                if !nonce_account::verify_nonce_account(&nonce_account, &durable_nonce)
+                    && working_bank.get_signature_status_slot(signature).is_none()
```
Comment (t-nelson) [inline core/src/send_transaction_service.rs:225]: Should we still bail on nonced TXs after some time?  Seems like one could cause trouble by broadcasting a bunch and backing up the queue

Comment (CriesofCarrots) [reply core/src/send_transaction_service.rs:225]: My initial thought is that's not much different than broadcasting a bunch of identical recent-blockhash transactions and backing up the queue. But I'm not opposed to some kind of timeout failsafe. I think that would be done by arbitrarily picking a `last_valid_slot` for durable-nonced txs (and then undoing the else-if set up in this PR).
Is 300 slots a reasonable timeout?

Comment (t-nelson) [reply core/src/send_transaction_service.rs:225]: > My initial thought is that's not much different than broadcasting a bunch of identical recent-blockhash transactions and backing up the queue.

My thinking is that the TXs could be maliciously constructed to conflict with one another, resulting in a financially bound total pool size, draining at one TX per block.  Where as recent-blockhash TXs would be limited to a pool size of `MAX_RECENT_BLOCKHASHES` or else it starts draining at `N` TX per block where `N = POOL_SIZE / MAX_RECENT_BLOCKHASHES`

> But I'm not opposed to some kind of timeout failsafe. I think that would be done by arbitrarily picking a last_valid_slot for durable-nonced txs (and then undoing the else-if set up in this PR).

Looks like that would do the trick!

> Is 300 slots a reasonable timeout?

300 seems like as good a starting point as any!

Comment (CriesofCarrots) [reply core/src/send_transaction_service.rs:236]: Implemented here: https://github.com/solana-labs/solana/pull/14308/commits/3470fb820bc059f2a24a9a24ce40b2c2eb444970

File: core/src/send_transaction_service.rs (hunk 1 above)
Comment (CriesofCarrots) [reply core/src/send_transaction_service.rs:240]: Added nonced-tx metric here: https://github.com/solana-labs/solana/pull/14308/commits/5994c0104785d8bb4a48acbc49693ca2841eea4f

**Review by t-nelson — APPROVED** [review]
Thanks for addressing!  LGTM!

**Pull request merged**